rfd/0xxx-partial-cache-healthiness.md

```diff
@@ -0,0 +1,318 @@
+
+`cache.Cache` implements the `Events` interface so that it could fan out events to other consumers that want to watch
+a subset of its event stream.
+
+When `NewWatcher()` is called on a Cache instance that's already in a healthy state - it knows which resource kinds
+it's actually receiving and which were rejected. Based on that, `NewWatcher()` can fail immediately
+if some of the requested resource kinds aren't available but `AllowRejectedKinds` isn't set. Otherwise, it can partially
+succeed and pass the list of unavailable resource kinds through `watch.KindsRejectedInFlight` all the way
+to `fanoutWatcher.init()` where it will generate an `OpInit` event with `WatchStatus` attached, listing the rejected
+kinds.
+
+When `NewWatcher()` is called on an unhealthy cache though, it's only possible to perform partial validation and make sure
+that Cache is at least configured to watch the requested kinds. In this situation Cache will have to return a `fanoutWatcher`
+that might fail later. When Cache receives its `OpInit` event and propagates that to fanout watchers, each will have to
+re-check if the request conditions are still met. Some might need to close with errors.
+
+
+#### Implementation details: client.streamWatcher
+
+`streamWatcher` allows watching events on the Auth Service over a GRPC connection. To add partial success mode
+to it, we'll need to make a small update to the API protocol and add a field that will be used to pass 
+the `AllowRejectedKinds` flag.
+
+```diff
+message Watch {
+   repeated WatchKind Kinds = 1;
++   bool AllowRejectedKinds = 2;
+}
+```
+
+On the server side this method is implemented by several `NewWatcher()` implementations wrapping each other and adding
+validations on each level. The last implementation in the chain is `Cache` which is already covered above.   
+
+
+#### Implementation details: local.watcher
+
+This one just follows the general rules:
+- checks each requested kind, keeping track of rejected ones
+- if some kinds were rejected and `AllowRejectedKinds` isn't set, return the first original validation error
+- otherwise, generates `OpInit` with `WatchStatus` resource attached and the list of rejected kinds on it
+
+
+### Generic readGuard[G]
+
+Read guard should now make the routing decision based on the resource kind that's being accessed.
+
+As suggested in the original issue [#21586](https://github.com/gravitational/teleport/issues/21586), there's a simple
+solution in which we would just pass the kind to `(*Cache).read()` and based on the cache health and whether the kind
+was rejected we'd get a `readGuard` routing all calls either to cache or to the API client. This option doesn't protect
+from programming errors where somebody would request a guard for one kind and then would call a method associated with
+another one.
+
+Another option suggested in that issue is to leverage generics to make `readGuard` only provide methods specific to
+the requested kind. Here's how that could look:
```

Comment:
I really like the generics proposal.  Looking it over, I do think impl might be complex enough that it'd be best to categorize this as a "nice to have", or "future work".

Reply:
I was able to move a couple collectors to this new interface during prototyping and it wasn't too bad. The worst part was that some of the kinds don't have an exactly corresponding interface in the `services` package, e.g.:

```go
package services

type RoleGetter interface {
	GetRole(ctx context.Context, name string) (types.Role, error)
}
```

This one is missing `GetRoles()` required for `Cache`, so I had to define a private `roleGetter` interface in collections.go.

Reply:
So you suggest to go with the simpler option from https://github.com/gravitational/teleport/issues/21586 for now?

Reply:
Just a gut reaction... point is just that I wouldn't consider generics a blocker since we have a reasonable fallback in `read(kind)`.  If you feel good about it, go for it!  But if it ends up feeling like a pain, I wouldn't invest too much energy into it at this stage.

My guess is that `ListResources` would be the trickiest to handle correctly in the context of generics (`ListResources` is _always_ causing trouble lol).  Maybe take a stab at applying the proposed generic strategy to `ListResources`... if it ends up being super difficult, that might be a good indicator that we should just go with `read(kind)` for now.

Reply:
Got it. Thanks for the heads up about `ListResources`!

Reply:
`ListResources` using a generic `readGuard` is shaping up to look something like this https://github.com/gravitational/teleport/commit/f93dbfb5e1d2f3157d56bd87581b93d92e546011
It's still mostly an exception because there's no corresponding collection.

I'll continue converting other collections for for now, but let me know if this seems over the top — I can cut it short and get back to the simple implementation.